runtime/src/main/java/org/corfudb/runtime/collections/StreamPollingTask.java

```diff
@@ -0,0 +1,108 @@
+class StreamPollingTask implements Runnable {
+
+    // A period of time in ms to sleep before next cycle when poller gets no new data changes.
+    private static final Duration IDLE_WAIT_TIME_MS = Duration.ofMillis(50);
+
+    // Total amount of time to wait for putting the polled data changes into buffer if it is full.
+    private static final Duration QUEUE_FULL_BLOCK_TIME_MS = Duration.ofMillis(2_000);
+
+    // The streaming manager that is in charge of listener subscriptions.
+    private final StreamingManager streamingManager;
+
+    // The corfu transaction stream to poll data changes from.
+    private final IStreamView txnStream;
+
+    // The subscription context associated with this task.
+    private final StreamSubscription subscription;
+
+    // The Thread pool for executing stream polling tasks.
+    private final ExecutorService pollingExecutor;
+
+    // Last address of the data successfully processed by the buffer.
+    private long lastReadAddress;
+
+    StreamPollingTask(StreamingManager streamingManager, long lastAddress,
+                      StreamSubscription subscription, ExecutorService executor) {
+        this.streamingManager = streamingManager;
+        this.subscription = subscription;
+        this.pollingExecutor = executor;
+        this.lastReadAddress = lastAddress;
+        this.txnStream = subscription.getTxnStream();
+    }
+
+    @Override
+    public void run() {
+        try {
+            pollTxnStream();
+        } catch (Throwable throwable) {
+            StreamListener listener = subscription.getListener();
+            log.error("Encountered exception {} during txn stream polling, listener: {}, " +
+                    "namespace: {}", throwable, listener, subscription.getNamespace());
+            streamingManager.unsubscribe(listener, false);
+            listener.onError(throwable);
+        }
+    }
+
+    /**
+     * Poll new data changes from the transaction stream and put into
+     * the subscription's buffer one by one.
+     */
+    private void pollTxnStream() throws Exception {
+        // If listener already unsubscribed, do not process or schedule again.
+        if (subscription.isStopped()) {
+            return;
+        }
+
+        // Seek to next address and poll transaction updates.
+        txnStream.seek(lastReadAddress + 1L);
+        List<ILogData> updates = txnStream.remainingAtMost(subscription.getStreamBufferSize());
+
+        // No new updates, take a short break and poll again.
+        if (updates.isEmpty()) {
+            TimeUnit.MILLISECONDS.sleep(IDLE_WAIT_TIME_MS.toMillis());
+            pollingExecutor.submit(this);
+            return;
+        }
+
+        // Insert polled updates to the subscription buffer, with a shared
+        // fixed amount of time waiting for buffer being not full.
+        long remainingBlockTime = QUEUE_FULL_BLOCK_TIME_MS.toNanos();
+        for (ILogData update : updates) {
+            if (subscription.isStopped()) {
```

Comment:
already checked this on L72

Reply:
Yea, just check again here because we synchronize it, so the stop reqeust could happen after the previous check.